### PR TITLE
Fix Windows build issues

### PR DESCRIPTION
**Pull request by fassisrosa**

### CODE REVIEW

File: main/midnight/src/ui/uicommandwindow.cpp [hunk 1]
```diff
@@ -343,7 +343,7 @@ void uicommandwindow::setupApproachText()
         recruitable_characters[0]);
 }
 
-void uicommandwindow::setCommandTextName(layoutid_t id, bool enabled, id_t characterId)
+void uicommandwindow::setCommandTextName(layoutid_t id, bool enabled, mxid characterId)
```
Comment (fassisrosa) [inline main/midnight/src/ui/uicommandwindow.cpp:346]: No id_t definition present. Followed 'recruitable_characters' definition (line 343) to find that mxid could be used.

File: main/midnight/src/ui/uicommandwindow.cpp [hunk 2]
```diff
@@ -369,7 +369,7 @@ void uicommandwindow::setCommandText(layoutid_t id, const std::string& value)
     }
 }
 
-void uicommandwindow::addCommandText(Node* node)
+void uicommandwindow::addCommandText(cocos2d::Node* node)
```
Comment (fassisrosa) [inline main/midnight/src/ui/uicommandwindow.cpp:372]: VS compiler complaining about ambiguous access (see: https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2385?f1url=%3FappId%3DDev16IDEF1%26l%3DEN-US%26k%3Dk(C2385)%26rd%3Dtrue&view=msvc-170). Disambiguation by use of cocos2d:: prefix.

File: main/midnight/src/ui/uibookmenu.cpp [hunk 3]
```diff
@@ -138,9 +138,13 @@ void uibookmenu::Show()
...
+    // get parent so we can send events to 
+    // it once we do remove child
+    uipanel* menuParent = parent;
+
+    // remove child, this sets parent to null
     parent->removeChild(this);
     
     // give the parent events back
-    parent->resumeEvents();
+    menuParent->resumeEvents();
```
Comment (fassisrosa) [inline main/midnight/src/ui/uibookmenu.cpp:149]: Was getting null pointer exception error here. line 146 ends up setting parent to null when removing child.

File: main/midnight/src/ui/uicommandwindow.cpp [hunk 4]
```diff
@@ -340,10 +340,10 @@ void uicommandwindow::setupApproachText()
     setCommandTextName(
         ID_APPROACH,
         location_flags.Is(lif_recruitchar),
-        recruitable_characters[0]);
+        recruitable_characters.size() > 0 ?  recruitable_characters[0] : 0);
```
Comment (fassisrosa) [inline main/midnight/src/ui/uicommandwindow.cpp:343]: Initialization was failing here. rectruitable_characters array was empty, access to zero entry failed.

File: main/midnight/src/tme/scenario/item_character.cpp [hunk 5]
```diff
@@ -1663,11 +1663,11 @@ namespace tme {
...
             VALIDATE_INFO_BLOCK(out,INFO_CHARACTER,defaultexport::character_t);
 
-        out->looking = looking;
+            out->looking = looking;
             out->time = Time();
-            out->longname = longname;
-            out->shortname = shortname;
-//            out->images =  images;
+            out->longname = std::string(longname);
+            out->shortname = std::string(shortname);
```
Comment (fassisrosa) [inline main/midnight/src/tme/scenario/item_character.cpp:1669]: Straight assignment was causing the corruption of shortname. Creating new std::string generates clean string.

File: main/midnight/src/ui/uicommandwindow.cpp (hunk 1 above)
Comment (IcemarkUK) [reply main/midnight/src/ui/uicommandwindow.cpp:346]: doh! My bad!!!!

File: main/midnight/src/ui/uicommandwindow.cpp (hunk 2 above)
Comment (IcemarkUK) [reply main/midnight/src/ui/uicommandwindow.cpp:372]: Could you try adding the Node in to the top of the uicommandwindow class. You'll see there is already a Widget and Layout there. That's the pattern that I "normally" use! :)

File: main/midnight/src/ui/uicommandwindow.cpp (hunk 4 above)
Comment (IcemarkUK) [reply main/midnight/src/ui/uicommandwindow.cpp:343]: Nice catch!

**Review by IcemarkUK — CHANGES_REQUESTED** [review]

File: main/midnight/src/ui/uibookmenu.cpp (hunk 3 above)
Comment (fassisrosa) [reply main/midnight/src/ui/uibookmenu.cpp:149]: Following calls, this ends up calling:
CCNode.cpp::removeChild

that one in turn calls:
CCNode.cpp::detachChild

And that one does:
   // set parent nil at the end
    child->setParent(nullptr);

File: main/midnight/src/tme/scenario/item_character.cpp (hunk 5 above)
Comment (fassisrosa) [reply main/midnight/src/tme/scenario/item_character.cpp:1669]: I don't think others are strings... enums, numbers, right?

File: main/midnight/src/ui/uicommandwindow.cpp (hunk 2 above)
Comment (fassisrosa) [reply main/midnight/src/ui/uicommandwindow.cpp:372]: Like it. Changed, will be in next push. Thanks!

**Review by IcemarkUK — APPROVED** [review]

**Comment by IcemarkUK** [comment]
Yes, there are a number of strings. They are just not accessed in the front end. 

It’s very odd that these are a problem, because out->short name and shortname are both std::string I can only presume the windows compiler is doing something funky here. To be safe I have pushed a commit where I have changed all the others.

Chris

**Comment by IcemarkUK** [comment]
I’ve just a change the does the equivalent to this in all areas. uicommandwindow, uipopup, uibookmenu, and uscompass

Could you just give it a once over an make sure the method I chose actually works on windows.

Chris

**Comment by fassisrosa** [comment]
> 

Seems to be working fine. Compiled successfully and played a bit (not a lot, just a bit to spot check) both LoM and DDR.

**Pull request merged**